mm2src/mm2_metrics/src/adapt.rs

```diff
@@ -0,0 +1,381 @@
+        }
+
+        let mut gauges = HashMap::new();
+        for (key, gauge) in self.registry.get_gauge_handles() {
+            let value = gauge.get_inner().load(Ordering::Acquire);
+            let (_name, labels) = key_to_parts(&key, None);
+            let entry = gauges
+                .entry(key.to_string())
+                .or_insert_with(HashMap::new)
+                .entry(labels)
+                .or_insert(value);
+            *entry = value;
+        }
+
+        let mut histograms = HashMap::new();
+        let histogram_handles = self.registry.get_histogram_handles();
+        for (key, histogram) in histogram_handles {
+            let (name, labels) = key_to_parts(&key, None);
+            let value = histogram.get_inner().data().iter().sum();
+            let entry = histograms
+                .entry(name)
+                .or_insert_with(HashMap::new)
+                .entry(labels)
+                .or_insert(value);
+            *entry = value;
+        }
+
+        Snapshot {
+            counters,
+            gauges,
+            histograms,
+        }
+    }
+
+    fn prepare_tag_metrics(&self) -> Vec<PreparedMetric> {
+        let mut output: Vec<PreparedMetric> = vec![];
+
+        for (key, counter) in self.registry.get_counter_handles().drain() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            let (name, _labels) = key_to_parts(&key, None);
+            let mut name_value_map = HashMap::new();
+            name_value_map.insert(ScopedString::Owned(name), Integer::Unsigned(value));
+            output.push(PreparedMetric {
+                tags: labels_to_tags(key.labels()),
+                message: name_value_map_to_message(&name_value_map),
+            });
+        }
+
+        for (key, gauge) in self.registry.get_gauge_handles().drain() {
+            let value = gauge.get_inner().load(Ordering::Acquire);
+            let (name, _labels) = key_to_parts(&key, None);
+            let mut name_value_map = HashMap::new();
+            name_value_map.insert(ScopedString::Owned(name), Integer::Unsigned(value));
+            output.push(PreparedMetric {
+                tags: labels_to_tags(key.labels()),
+                message: name_value_map_to_message(&name_value_map),
+            });
+        }
+
+        for (key, histogram) in self.registry.get_histogram_handles().drain() {
+            let value: f64 = histogram.get_inner().data().iter().sum();
+            let (name, _labels) = key_to_parts(&key, None);
+            let mut name_value_map = HashMap::new();
+
+            name_value_map.insert(ScopedString::Owned(name), Integer::Unsigned(value as u64));
+            output.push(PreparedMetric {
+                tags: labels_to_tags(key.labels()),
+                message: name_value_map_to_message(&name_value_map),
+            });
+        }
+
+        output
+    }
+}
+
+#[derive(Clone)]
+pub struct MmRecorder {
+    pub(crate) inner: Arc<Mutex<Inner>>,
```

Comment:
Could you please change the names like it's done now:
There is `MetricsArc` and `MetricsWeak`. In your case `MetricsArc` is `MmRecorder`.
There is `Metrics`, where `MetricsArc(Arc<Mutex<Metrics>>)`. But I like the `MmRecorder` name instead.
So in your case `Inner` should be called `MmRecorder`.
And then you can implement `metrics::Recorder` for `MmRecorder`.

What do you think?